### PR TITLE
Add IGNORE NULLS clause to LAG/LEAD functions

### PR DESCRIPTION
**Pull request by ptkool**
See issue https://github.com/prestodb/presto/issues/4554.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/window/WindowAnnotationsParser.java [hunk 1]
```diff
@@ -62,6 +62,8 @@ private static SqlWindowFunction parse(Class<? extends WindowFunction> clazz, Wi
                 argumentTypes,
                 false);
 
-        return new SqlWindowFunction(new ReflectionWindowFunctionSupplier<>(signature, clazz));
+        boolean canIgnoreNulls = window.name().equalsIgnoreCase("lag") || window.name().equalsIgnoreCase("lead");
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/window/WindowAnnotationsParser.java]: Instead of hard-coding it for these two, we should discover it from function definitions so that it can apply to any function that declares they support this option (e.g., via an annotation)

File: presto-spi/src/main/java/com/facebook/presto/spi/function/WindowFunctionOptions.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+@Retention(RUNTIME)
+@Target(TYPE)
+public  @interface WindowFunctionOptions
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/WindowFunctionOptions.java]: Formatting: remove the extra space after `public`

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 3]
```diff
@@ -281,7 +281,8 @@ primaryExpression
     | '(' expression (',' expression)+ ')'                                                #rowConstructor
     | ROW '(' expression (',' expression)* ')'                                            #rowConstructor
     | qualifiedName '(' ASTERISK ')' filter? over?                                        #functionCall
-    | qualifiedName '(' (setQuantifier? expression (',' expression)*)? ')' filter? over?  #functionCall
+    | qualifiedName '(' (setQuantifier? expression (',' expression)*)? ')' nullTreatment?
```
Comment (electrum) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: Put the `nullTreatment` on the next line, too, so all the options are together

File: presto-main/src/main/java/com/facebook/presto/operator/window/WindowAnnotationsParser.java [hunk 4]
```diff
@@ -37,12 +38,15 @@ private WindowAnnotationsParser() {}
...
+        boolean canIgnoreNulls = options.length > 0 && options[0].canIgnoreNulls();
         return Stream.of(signatures)
-                .map(signature -> parse(clazz, signature))
+                .map(signature -> parse(clazz, signature, canIgnoreNulls))
                 .collect(toImmutableList());
     }
 
-    private static SqlWindowFunction parse(Class<? extends WindowFunction> clazz, WindowFunctionSignature window)
+    private static SqlWindowFunction parse(Class<? extends WindowFunction> clazz, WindowFunctionSignature window,
+                                           boolean canIgnoreNulls)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/window/WindowAnnotationsParser.java]: See https://github.com/airlift/codestyle

We either wrap all arguments or none. If wrapped, they should use a continuation indent of 8 spaces (not aligned to parentheses).

File: presto-main/src/main/java/com/facebook/presto/operator/window/WindowAnnotationsParser.java [hunk 5]
```diff
@@ -37,12 +38,15 @@ private WindowAnnotationsParser() {}
     {
         WindowFunctionSignature[] signatures = clazz.getAnnotationsByType(WindowFunctionSignature.class);
         checkArgument(signatures.length > 0, "Class is not annotated with @WindowFunctionSignature: %s", clazz.getName());
+        WindowFunctionOptions[] options = clazz.getAnnotationsByType(WindowFunctionOptions.class);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/window/WindowAnnotationsParser.java]: This should `getAnnotations()` since there can be at most one instance of the annotation (it's not repeatable)

File: presto-main/src/main/java/com/facebook/presto/operator/window/LagFunction.java [hunk 6]
```diff
@@ -52,14 +57,22 @@ public void processRow(BlockBuilder output, int frameStart, int frameEnd, int cu
...
-            }
-            else if (defaultChannel >= 0) {
-                windowIndex.appendTo(defaultChannel, currentPosition, output);
-            }
-            else {
-                output.appendNull();
+            while (true) {
+                if ((valuePosition >= 0) && (valuePosition <= currentPosition)) {
+                    if (ignoreNulls && windowIndex.isNull(valueChannel, Ints.checkedCast(valuePosition))) {
+                        --valuePosition;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/window/LagFunction.java]: Use `valuePosition--`

File: presto-main/src/main/java/com/facebook/presto/operator/window/LagFunction.java [hunk 7]
```diff
@@ -52,14 +57,22 @@ public void processRow(BlockBuilder output, int frameStart, int frameEnd, int cu
...
 
-            if ((valuePosition >= 0) && (valuePosition <= currentPosition)) {
-                windowIndex.appendTo(valueChannel, Ints.checkedCast(valuePosition), output);
-            }
-            else if (defaultChannel >= 0) {
-                windowIndex.appendTo(defaultChannel, currentPosition, output);
-            }
-            else {
-                output.appendNull();
+            while (true) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/window/LagFunction.java]: This control flow is hard to follow. Break out the null skipping logic to happen before the append logic.

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 8]
```diff
@@ -1165,6 +1165,14 @@ public Node visitFunctionCall(SqlBaseParser.FunctionCallContext context)
 
         QualifiedName name = getQualifiedName(context.qualifiedName());
 
+        boolean ignoreNulls = false;
+        if (context.nullTreatment() != null) {
+            String functionName = name.toString();
+            check(functionName.equalsIgnoreCase("lag") || functionName.equalsIgnoreCase("lead"),
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: This check needs to happen in the analyzer. It should throw `SemanticException(NOT_SUPPORTED)`

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 9]
```diff
@@ -612,6 +618,8 @@ CALL: 'CALL';
 PREPARE: 'PREPARE';
 DEALLOCATE: 'DEALLOCATE';
 EXECUTE: 'EXECUTE';
+IGNORE: 'IGNORE';
```
Comment (electrum) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: Make these non-reserved words by adding them to the `nonReserved` list above.

File: presto-main/src/main/java/com/facebook/presto/operator/window/ReflectionWindowFunctionSupplier.java [hunk 10]
```diff
@@ -32,35 +32,49 @@
...
+    private final boolean canIgnoreNulls;
 
     public ReflectionWindowFunctionSupplier(String name, Type returnType, List<? extends Type> argumentTypes, Class<T> type)
     {
-        this(new Signature(name, WINDOW, returnType.getTypeSignature(), Lists.transform(argumentTypes, Type::getTypeSignature)), type);
+        this(new Signature(name, WINDOW, returnType.getTypeSignature(), Lists.transform(argumentTypes, Type::getTypeSignature)), false, type);
     }
 
-    public ReflectionWindowFunctionSupplier(Signature signature, Class<T> type)
+    public ReflectionWindowFunctionSupplier(String name, Type returnType, boolean canIgnoreNulls, List<? extends Type> argumentTypes, Class<T> type)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/window/ReflectionWindowFunctionSupplier.java]: The constructor above is only used for testing, so add another parameter to it rather than adding another constructor.

File: presto-main/src/main/java/com/facebook/presto/operator/window/ReflectionWindowFunctionSupplier.java [hunk 11]
```diff
@@ -32,35 +32,49 @@
...
                 constructor = type.getConstructor();
             }
+            else if (canIgnoreNulls) {
+                constructor = type.getConstructor(List.class, boolean.class);
+            }
             else {
                 constructor = type.getConstructor(List.class);
             }
+
+            this.canIgnoreNulls = canIgnoreNulls;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/window/ReflectionWindowFunctionSupplier.java]: Put this line right after the `super()` call.

File: presto-main/src/main/java/com/facebook/presto/operator/window/LeadFunction.java [hunk 12]
```diff
@@ -52,14 +57,22 @@ public void processRow(BlockBuilder output, int frameStart, int frameEnd, int cu
...
 
-            if ((valuePosition >= 0) && (valuePosition < windowIndex.size())) {
-                windowIndex.appendTo(valueChannel, Ints.checkedCast(valuePosition), output);
-            }
-            else if (defaultChannel >= 0) {
-                windowIndex.appendTo(defaultChannel, currentPosition, output);
-            }
-            else {
-                output.appendNull();
+            while (true) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/window/LeadFunction.java]: Same comments as `lag`

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 13]
```diff
@@ -1165,6 +1165,14 @@ public Node visitFunctionCall(SqlBaseParser.FunctionCallContext context)
...
+                    "IGNORE NULLS clause not valid for '" + functionName + "' function", context);
+            ignoreNulls = context.nullTreatment().getText().equalsIgnoreCase("ignorenulls");
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: This should use the `Token` type rather than string comparison. For example, see how window frame bounds are handled.

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestStatementBuilder.java [hunk 14]
```diff
@@ -95,6 +95,11 @@ public void testStatementBuilder()
                 "from emp");
 
         printStatement("" +
+                "select depname, empno, salary\n" +
+                ", lag(salary, 1) ignore nulls over (partition by depname)\n" +
```
Comment (electrum) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestStatementBuilder.java]: Move this line to the above `printStatement` call that has all the other window functions. No need to introduce a new query. Also add a test for `respect nulls`.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/FunctionCall.java [hunk 15]
```diff
@@ -26,44 +26,60 @@
...
+        this(Optional.of(location), name, Optional.empty(), Optional.empty(), distinct, false, arguments);
     }
 
     public FunctionCall(QualifiedName name, Optional<Window> window, boolean distinct, List<Expression> arguments)
     {
-        this(Optional.empty(), name, window, Optional.empty(), distinct, arguments);
+        this(Optional.empty(), name, window, Optional.empty(), distinct, false, arguments);
+    }
+
+    public FunctionCall(QualifiedName name, Optional<Window> window, boolean distinct, boolean ignoreNulls, List<Expression> arguments)
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/FunctionCall.java]: Where are all these new constructors used? This class already has too many, so I'd prefer to simply add as few as possible (or none).

I think we should add it as a parameter to all of the constructors that take a `Window` and update all the call sites to make sure it's passed through correctly. We have a lot of places that rewrite the AST, and it's easy for one of them to miss the parameter and silently change the behavior (thus producing wrong answers).

**Comment by ptkool** [comment]
Please review latest changes. @electrum @martint

**Comment by ptkool** [comment]
@electrum Any more thoughts on this?

File: presto-docs/src/main/sphinx/functions/window.rst [hunk 16]
```diff
@@ -106,18 +106,26 @@ Value Functions
     the window, null is returned.  It is an error for the offset to be zero or
     negative.
 
-.. function:: lead(x[, offset [, default_value]]) -> [same as input]
+.. function:: lead(x[, offset [, default_value]]) ignore nulls -> [same as input]
```
Comment (maciejgrzybek) [inline presto-docs/src/main/sphinx/functions/window.rst]: Please place `ignore nulls` in square brackets because it's optional.

File: presto-docs/src/main/sphinx/functions/window.rst [hunk 17]
```diff
@@ -106,18 +106,26 @@ Value Functions
...
     offset can be any scalar expression.  The default ``offset`` is ``1``. If the
     offset is null or larger than the window, the ``default_value`` is returned,
     or if it is not specified ``null`` is returned.
 
-.. function:: lag(x[, offset [, default_value]]) -> [same as input]
+    If ``IGNORE NULLS`` is specified, all rows where ``x`` is null are excluded from the calculation. If ``IGNORE NULLS``
+    is specified and ``x`` is null for all rows, the ``default_value`` is returned. If ``IGNORE NULLS`` is
+    not specified, null values are respected.
+
+.. function:: lag(x[, offset [, default_value]]) ignore nulls -> [same as input]
```
Comment (maciejgrzybek) [inline presto-docs/src/main/sphinx/functions/window.rst]: ditto

File: presto-main/src/main/java/com/facebook/presto/operator/window/AggregateWindowFunction.java [hunk 18]
```diff
@@ -106,7 +106,7 @@ public static WindowFunctionSupplier supplier(Signature signature, final Interna
         return new AbstractWindowFunctionSupplier(signature, null)
         {
             @Override
-            protected WindowFunction newWindowFunction(List<Integer> inputs)
+            protected WindowFunction newWindowFunction(List<Integer> inputs, boolean ignoreNulls)
             {
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/operator/window/AggregateWindowFunction.java]: Add `checkArgument(ignoreNulls == false)` to ensure that no one uses `ignore nulls` for not (yet) supported functions, e.g. `first_value`.

File: presto-main/src/test/java/com/facebook/presto/operator/TestWindowOperator.java [hunk 19]
```diff
@@ -80,11 +80,11 @@
     );
 
     private static final List<WindowFunctionDefinition> LAG = ImmutableList.of(
-            window(new ReflectionWindowFunctionSupplier<>("lag", VARCHAR, ImmutableList.of(VARCHAR, BIGINT, VARCHAR), LagFunction.class), VARCHAR, UNBOUNDED_FRAME, 1, 3, 4)
+            window(new ReflectionWindowFunctionSupplier<>("lag", VARCHAR, true, ImmutableList.of(VARCHAR, BIGINT, VARCHAR), LagFunction.class), VARCHAR, UNBOUNDED_FRAME, 1, 3, 4)
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/operator/TestWindowOperator.java]: By default value WFs are `RESPECT NULLS` so please don't change that test to non-default behavior. Add new tests for non-default behavior to cover your newly added feature, instead.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionListBuilder.java [hunk 20]
```diff
@@ -60,7 +61,10 @@ public FunctionListBuilder window(String name, Class<? extends ValueWindowFuncti
                 parseTypeSignature(typeVariable),
                 Arrays.asList(argumentTypes).stream().map(TypeSignature::parseTypeSignature).collect(toImmutableList()),
                 false);
-        boolean canIgnoreNulls = name.equalsIgnoreCase("lag") || name.equalsIgnoreCase("lead");
+
+        WindowFunctionOptions[] options = clazz.getAnnotationsByType(WindowFunctionOptions.class);
+        boolean canIgnoreNulls = options.length > 0 && options[0].canIgnoreNulls();
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionListBuilder.java]: Please add `checkState(options.length == 0 || options.length == 1)` because you don't want to support multiple annotations of `WindowFunctionOptions` here. Even if you wanted to, you'd need different `canIgnoreNulls` logic.

File: presto-main/src/main/java/com/facebook/presto/operator/window/WindowAnnotationsParser.java (hunk 5 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/operator/window/WindowAnnotationsParser.java]: Isn't it better to use `getAnnotationsByType` and assert that we expect only one annotation here, like I suggested before?
By simply using `getAnnotations` here instead, we don't get rid of the problem of possibly having multiple instances of the same annotation. Instead, we're dealing with a problem of having annotations of other types too.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 21]
```diff
@@ -709,7 +709,7 @@ public PhysicalOperation visitWindow(WindowNode node, LocalExecutionPlanContext
                 Symbol symbol = entry.getKey();
                 WindowFunctionSupplier windowFunctionSupplier = metadata.getFunctionRegistry().getWindowFunctionImplementation(signature);
                 Type type = metadata.getType(signature.getReturnType());
-                windowFunctionsBuilder.add(window(windowFunctionSupplier, type, functionCall.isIgnoreNulls(), arguments.build()));
+                windowFunctionsBuilder.add(window(windowFunctionSupplier, type, frameInfo, functionCall.isIgnoreNulls(), arguments.build()));
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: This seems like not belonging to this commit.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/ExpressionTreeRewriter.java [hunk 22]
```diff
@@ -549,7 +549,7 @@ public Expression visitFunctionCall(FunctionCall node, Context<C> context)
 
             if (!sameElements(node.getArguments(), arguments.build()) || !sameElements(rewrittenWindow, node.getWindow())
                     || !sameElements(filter, node.getFilter())) {
-                return new FunctionCall(node.getName(), rewrittenWindow, node.isDistinct(), node.isIgnoreNulls(), arguments.build());
+                return new FunctionCall(node.getName(), rewrittenWindow, filter, node.isDistinct(), node.isIgnoreNulls(), arguments.build());
```
Comment (maciejgrzybek) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/ExpressionTreeRewriter.java]: Ditto, it seems not belonging to this commit.

File: presto-main/src/test/java/com/facebook/presto/operator/TestWindowOperator.java [hunk 23]
```diff
@@ -65,27 +65,27 @@
     private static final FrameInfo UNBOUNDED_FRAME = new FrameInfo(RANGE, UNBOUNDED_PRECEDING, Optional.empty(), UNBOUNDED_FOLLOWING, Optional.empty());
 
     private static final List<WindowFunctionDefinition> ROW_NUMBER = ImmutableList.of(
-            window(new ReflectionWindowFunctionSupplier<>("row_number", BIGINT, ImmutableList.of(), RowNumberFunction.class), BIGINT, UNBOUNDED_FRAME)
+            window(new ReflectionWindowFunctionSupplier<>("row_number", BIGINT, false, ImmutableList.<Type>of(), RowNumberFunction.class), BIGINT, UNBOUNDED_FRAME)
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/operator/TestWindowOperator.java]: Explicit generic argument should not be necessary here.

File: presto-main/src/test/java/com/facebook/presto/operator/window/TestLagFunction.java [hunk 24]
```diff
@@ -190,5 +191,33 @@ public void testLagFunction()
...
+                        .row(5L, "F", "F")
+                        .row(6L, "F", "F")
+                        .row(7L, null, "F")
+                        .row(34L, "O", "F")
+                        .row(null, "F", "O")
+                        .row(null, "O", "F")
+                        .row(null, null, "O")
+                        .row(null, null, "O")
+                        .build());
     }
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/operator/window/TestLagFunction.java]: Please add tests with `RESPECT NULLS` explicitly set, in addition to `IGNORE NULLS` tests.

File: presto-main/src/main/java/com/facebook/presto/operator/window/WindowAnnotationsParser.java [hunk 25]
```diff
@@ -37,12 +38,15 @@ private WindowAnnotationsParser() {}
...
+        boolean canIgnoreNulls = options != null && options.canIgnoreNulls();
         return Stream.of(signatures)
-                .map(signature -> parse(clazz, signature))
+                .map(signature -> parse(clazz, signature, canIgnoreNulls))
                 .collect(toImmutableList());
     }
 
-    private static SqlWindowFunction parse(Class<? extends WindowFunction> clazz, WindowFunctionSignature window)
+    private static SqlWindowFunction parse(Class<? extends WindowFunction> clazz, WindowFunctionSignature window,
+            boolean canIgnoreNulls)
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/operator/window/WindowAnnotationsParser.java]: Please put all arguments in one line or each in separate line.

File: presto-spi/src/main/java/com/facebook/presto/spi/function/WindowFunctionOptions.java [hunk 26]
```diff
@@ -0,0 +1,27 @@
...
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+@Retention(RUNTIME)
+@Target(TYPE)
+public @interface WindowFunctionOptions
```
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/WindowFunctionOptions.java]: I think if you implemented `ignore nulls` for three remaining `ValueWindowFunctions` (`FirstValue`, `LastValue` and `NthValue`), then we'd not need this annotation but instead we could distinguish `canIgnoreNulls`-functions by interface they implement (`canIgnoreNulls = instanceof ValueWindowFunctions`).

File: presto-main/src/main/java/com/facebook/presto/operator/window/WindowAnnotationsParser.java (hunk 25 above)
Comment (ptkool) [reply presto-main/src/main/java/com/facebook/presto/operator/window/WindowAnnotationsParser.java]: Ok.

File: presto-spi/src/main/java/com/facebook/presto/spi/function/WindowFunctionOptions.java (hunk 26 above)
Comment (ptkool) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/WindowFunctionOptions.java]: I will do this.

**Comment by tdowns** [comment]
@ptkool is this still an open request?

**Comment by ptkool** [comment]
@tdowns Yes, it is. Should have time this week to finish it.

**Comment by joshpeng** [comment]
@ptkool Really appreciate your work. Was wondering if there was any status update on this? Thanks.

**Comment by ptkool** [comment]
@joshpeng Just need to finish up changes to `NthValue` function.

**Comment by MichaelChirico** [comment]
@ptkool offering some encouragement. you got this!

**Comment by Lunatictwo** [comment]
@ptkool how is the process of this PR?

**Comment by ptkool** [comment]
@Lunatictwo I'm sorry, I don't know what you're asking.
